Index: evolution-data-server-2.27.2/camel/camel.c
===================================================================
--- evolution-data-server-2.27.2.orig/camel/camel.c
+++ evolution-data-server-2.27.2/camel/camel.c
@@ -38,6 +38,8 @@
 #include <glib.h>
 #include <glib/gi18n-lib.h>
 
+#include <nss-shared-helper.h>
+
 #include "camel.h"
 #include "camel-certdb.h"
 #include "camel-debug.h"
@@ -94,17 +96,21 @@ camel_init (const char *configdir, gbool
 		nss_configdir = g_win32_locale_filename_from_utf8 (configdir);
 #endif
 
-		if (!NSS_IsInitialized()) {
-			nss_initialized = 1;
+		if (nsshelp_open_db ("evolution", nss_configdir, 0) != SECSuccess) {
+			g_warning ("Opening non-shared cert database");
+
+			if (!NSS_IsInitialized()) {
+				nss_initialized = 1;
 
-			if (NSS_InitReadWrite (nss_configdir) == SECFailure) {
-				/* fall back on using volatile dbs? */
-				if (NSS_NoDB_Init (nss_configdir) == SECFailure) {
-					g_free (nss_configdir);
-					g_warning ("Failed to initialize NSS");
-					nss_initialized = 0;
-					PR_Unlock(nss_initlock);
-					return -1;
+				if (NSS_InitReadWrite (nss_configdir) == SECFailure) {
+					/* fall back on using volatile dbs? */
+					if (NSS_NoDB_Init (nss_configdir) == SECFailure) {
+						g_free (nss_configdir);
+						g_warning ("Failed to initialize NSS");
+						nss_initialized = 0;
+						PR_Unlock(nss_initlock);
+						return -1;
+					}
 				}
 			}
 		}
Index: evolution-data-server-2.27.2/configure.ac
===================================================================
--- evolution-data-server-2.27.2.orig/configure.ac
+++ evolution-data-server-2.27.2/configure.ac
@@ -575,6 +575,7 @@ if test "x${enable_nss}" = "xyes" || tes
 		done
 		
 		if test -n "$mozilla_nspr" -a -n "$mozilla_nss"; then
+			mozilla_nss="$mozilla_nss nss-shared-helper"
 			msg_ssl="yes (Mozilla NSS)"
 			if test "x$enable_smime" = "xyes"; then
 				AC_DEFINE(ENABLE_SMIME,1,[Define if SMIME should be enabled])
