Index: e-util/e-gui-utils.c
===================================================================
--- e-util/e-gui-utils.c	(revision 36925)
+++ e-util/e-gui-utils.c	(working copy)
@@ -64,3 +64,30 @@
 
 	return pixbuf;
 }
+
+GtkWidget *e_create_image_widget(gchar *name, 	 
+                                  gchar *string1, gchar *string2, 	 
+                                  gint int1, gint int2) 	 
+{ 	 
+         GtkWidget *alignment = NULL; 	 
+         GtkWidget *w; 	 
+  	 
+         if (string1) { 	 
+                 w = gtk_image_new_from_icon_name ( 	 
+                         string1, GTK_ICON_SIZE_DIALOG); 	 
+  	 
+                 gtk_misc_set_alignment (GTK_MISC (w), 0.5, 0.5); 	 
+  	 
+                 alignment = gtk_widget_new(gtk_alignment_get_type(), 	 
+                                            "child", w, 	 
+                                            "xalign", (double) 0, 	 
+                                            "yalign", (double) 0, 	 
+                                            "xscale", (double) 0, 	 
+                                            "yscale", (double) 0, 	 
+                                            NULL); 	 
+  	 
+                 gtk_widget_show_all (alignment); 	 
+         } 	 
+  	 
+         return alignment; 	 
+}
