Index: addressbook/printing/e-contact-print.c
===================================================================
--- addressbook/printing/e-contact-print.c	(revision 36615)
+++ addressbook/printing/e-contact-print.c	(working copy)
@@ -266,8 +266,20 @@
 		if (value == NULL || *value == '\0')
 			continue;
 
-		text = g_strdup_printf ("%s:  %s",
-			e_contact_pretty_name (field), value);
+		if (field == E_CONTACT_BOOK_URI) {
+			gchar **tmpv;
+			tmpv = g_strsplit (value, ";", 2);
+			/* To translators: While printing contacts display the addressbook
+			   to which the contact belongs.
+			   Example: "Address Book:  Frequent Contacts"
+			*/
+			text = g_strdup_printf ("%s:  %s",
+				_( "Address Book" ), tmpv[1]);
+			g_strfreev (tmpv);
+		}
+		else
+			text = g_strdup_printf ("%s:  %s",
+				e_contact_pretty_name (field), value);
 
 		if (ctxt->pages == ctxt->page_nr)
 			e_contact_output (
