diff --git a/gswitchit/gswitchit-applet.c b/gswitchit/gswitchit-applet.c
index dab7ebf..96e9a55 100644
--- a/gswitchit/gswitchit-applet.c
+++ b/gswitchit/gswitchit-applet.c
@@ -448,7 +448,7 @@ GSwitchItAppletCmdHelp (BonoboUIComponent
 			* uic, GSwitchItApplet * sia, const gchar * verb)
 {
 	gtk_show_uri (gtk_widget_get_screen (GTK_WIDGET (sia->applet)),
-		"ghelp:gswitchit-view",
+		"ghelp:gswitchit",
 		gtk_get_current_event_time (),
 		NULL);
 } void
