 src/Makefile.am |    3 ---
 1 files changed, 0 insertions(+), 3 deletions(-)

diff --git a/src/Makefile.am b/src/Makefile.am
index d043432..510b472 100755
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -44,8 +44,6 @@ cairo_dock_CFLAGS = \
 	-DCAIRO_DOCK_THEMES_DIR=\""$(CAIRO_DOCK_THEMES_DIR)"\"\
 	-DGL_GLEXT_PROTOTYPES=\""1"\"\
 	-std=c99\
-	-O3\
-	-g -ggdb\
 	-Wstrict-prototypes -Wextra -Wwrite-strings -Wuninitialized -Werror-implicit-function-declaration
 #-Wunreachable-code -Wno-unused-parameter -Wall 
 
@@ -183,7 +181,6 @@ cairo_dock_SOURCES = \
 	cairo-dock-task.c\
 	cairo-dock-task.h\
 	texture-gradation.h
-	
 
 cairo_dockincludedir =\
 	$(includedir)/cairo-dock
